Expose color chooser widgets

ericmehl opened the pull request:
This reveals the new widgets added in and #5962, #5944 and #5999 and gives control over their visibility.

There's probably some discussion to have over the API introduced in f0258ed8ff5ef485077b3b006fe09f51c4cd0562 for getting and setting widget visibility in the Color Chooser, and how the state is stored.

When creating a color chooser, there are three places it's looking to see if there are settings for widget visibility :
1. The `Color3f` / `Color4f` plug(s) the widget is associated with
2. `colorChooser:inlineOptions` / `colorChooser:dialogueOptions` `userDefault` global metadata entry
3. A per-session `colorChosoer:inlineOptions` / `colorChooser:dialogueOptions` `sessionDefault` global metadata entry

This way you get the state associated with the plug if there is one. If there's no opinion for the plug, we prioritize `userDefault` on the assumption that someone went to the trouble to set that in a startup script and we should honor that. If that does not exist, we use a per-session variable on the assumption that if a user turned off some elements for one plug's widget, there's a good chance they will want them off for other plugs as well.

And lastly, there are separate options metadata entries for the inline and dialogue options. Mixing them seems like the logic and expectations could become unclear, and my theory is that people tend to use the inline vs dialogue for different reasons and may want different UI elements visible in each.

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

ericmehl commented:
I made some changes today and squashed them into the original commits since they haven't been reviewed yet. They're cosmetic for the most part - tweaked some icons / indicators, added text entries to the menu for the color field component choices and made some code improvements.

Ready for review.

johnhaddon commented on python/GafferUI/ColorChooser.py:
```diff
@@ -611,6 +614,31 @@ def setErrored( self, errored ) :
+
+		return "".join( [ k for k, v in self.__sliders.items() if v.getVisible() ] )
+
+	def setStaticComponent( self, component ) :
+
+		self.__setStaticComponentInternal( component )
+
+	def getStaticComponent( self ) :
+
+		c, staticComponent = self.__colorField.getColor()
+		return staticComponent
+
+	def setColorFieldVisible( self, visible ) :
+
+		self.__setColorFieldVisibleInternal( visible )
+
+	def getColorFieldVisible( self ) :
+
+		return self.__colorField.getVisible()
```
I did wonder if we could have just one property here instead of two, and use `None` to mean invisible. But maybe that doesn't work if we want the menu to have separate options for visibility and component...

johnhaddon commented on python/GafferUI/ColorChooser.py:
```diff
@@ -611,6 +614,31 @@ def setErrored( self, errored ) :
+
+	def setStaticComponent( self, component ) :
+
+		self.__setStaticComponentInternal( component )
```
Do we need a separate internal function? Could the implementation just move here?

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/6003/commits/591819a5ff5d6ee70dfb7a4297375321b837fa0d

johnhaddon commented on python/GafferUI/ColorChooser.py:
```diff
@@ -611,6 +614,31 @@ def setErrored( self, errored ) :
+
+		return "".join( [ k for k, v in self.__sliders.items() if v.getVisible() ] )
+
+	def setStaticComponent( self, component ) :
```
Is `StaticComponent` meaningful enough in the context of ColorChooser? Should it be `colorFieldStaticComponent()`?

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/6003/commits/680a97c9d5bdd91088b7ab19e7d07d9c93d55040

johnhaddon commented on python/GafferUI/ColorChooser.py:
```diff
@@ -620,6 +648,29 @@ def colorChangedSignal( self ) :
 
 		return self.__colorChangedSignal
 
+	## A signal emitted whenver the visible components are changed. Slots
```
whenver -> whenever

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/6003/commits/9ec28a6888470afd72b38d584d3b44a8b564b69b

johnhaddon commented on python/GafferUI/ColorChooser.py:
```diff
@@ -620,6 +648,29 @@ def colorChangedSignal( self ) :
 
+	## A signal emitted whenver the visible components are changed. Slots
+	# should have the signature slot( ColorChooser, visibleComponents ).
+	# `visibleComponents` is a string representing the components currently
+	# visible.
+	def visibleComponentsChangedSignal( self ) :
+
+		return self.__visibleComponentsChangedSignal
+
+	## A signal emitted whenver the static component is changed. Slots
+	# should have the signature slot( ColorChooser, staticComponent ).
+	# `staticComponent` is a single character string representing the
+	# current static component.
+	def staticComponentChangedSignal( self ) :
+
+		return self.__staticComponentChangedSignal
+
+	## A signal emitted whenever the visibility of the color field changes.
+	# Slots should have the signature slot( ColorChooser, visible ).
+	# `visible` is a boolean representing the current visibility.
```
I wonder how useful the property (`staticComponent` etc) arguments to the slots are, given that the property can always be queried directly from the ColorChooser.

ericmehl replied:
Yeah, not much use in those extra arguments. Fixed in https://github.com/GafferHQ/gaffer/pull/6003/commits/7803fb8c31eb830efe67507b45d926e237015fb0

johnhaddon commented on python/GafferUI/ColorChooserPlugValueWidget.py:
```diff
@@ -89,6 +126,34 @@ def __colorChanged( self, colorChooser, reason ) :
+
+		v = sole( Gaffer.Metadata.value( p, "colorChooser:inlineOptions" ) for p in self.getPlugs() )
+		if v is None :
+			v  = Gaffer.Metadata.value( "colorChooser:inlineOptions", "userDefault" )
```
`userDefault` has quite a specific meaning already - a default value assigned to plugs on newly-created nodes - and I don't think the semantics match what we're doing here.

I think this is redundant anyway though - we can achieve the same thing using the existing metadata inheritance rules. Registering against `Color3fPlug` and `Color4fPlug` gives us the fallback behaviour, and we can also use node-specific registrations to provide an in-between level of control.

ericmehl replied:
I did a big overhaul of how this works for this and the following comment. It does feel a lot more streamlined now. We're only reading from the plug value, which will fall back to `Color3fPlug`, `Color4fPlug` if the plug instance doesn't have an opinion. Done in https://github.com/GafferHQ/gaffer/pull/6003/commits/8bf9543bafcf26dd41e265f7f18376a2d8460a4a

johnhaddon commented on python/GafferUI/ColorChooserPlugValueWidget.py:
```diff
@@ -89,6 +126,34 @@ def __colorChanged( self, colorChooser, reason ) :
+		if v is None :
+			v  = Gaffer.Metadata.value( "colorChooser:inlineOptions", "userDefault" )
+			if v is None :
+				v = Gaffer.Metadata.value( "colorChooser:inlineOptions", "sessionDefault" ) or IECore.CompoundData()
```
I'm not sure what this provides on top of the usual metadata inheritance rules either - is there something I'm missing?

Taking on board recent comments about startup files being a bit much for a non-technical standalone user, I wonder if we should add a "Save Defaults" menu item that creates `~/gaffer/startup/gui/colorChooser.py` and makes the `Color[34]fPlug` registrations based on the values from the current plug?

ericmehl replied:
My goal was to treat `userDefault` (I see your point on registering to `Color[3,4]fPlug` so I'll refer to that from now on) as the behavior the user wants regardless of what they change during a session. I figured registering metadata in a startup script is a very proactive choice and we should stick to that throughout a session, regardless of what a user changes during a session.

To accomplish that I needed the two separate metadata entries so there's one entry that gets updated as the user makes choices (unless there's a setting from a startup script) and (possibly) one that is kept static because it was set in a startup script.

But maybe that behavior is not the best, or too complicated? If we always register a choice to a `Color3fPlug`, users can set options they get at startup, and then if they change an option, that will become the new default for `ColorChooser` without a plug-specific registration.

ericmehl replied:
This was improved in the great Color Chooser metadata overhaul of 2024: https://github.com/GafferHQ/gaffer/pull/6003/commits/8bf9543bafcf26dd41e265f7f18376a2d8460a4a. Now we have an explicit method of saving defaults, including creating a startup script to load the settings.

johnhaddon commented on python/GafferUI/ColorSwatchPlugValueWidget.py:
```diff
@@ -214,3 +252,34 @@ def __buttonClicked( self, button ) :
+		if Gaffer.Metadata.value( "colorChooser:dialogueOptions", "userDefault" ) is None :
+			sessionOptions = Gaffer.Metadata.value( "colorChooser:dialogueOptions", "sessionDefault" )
+			if sessionOptions is None :
+				sessionOptions = IECore.CompoundData()
```
Smooshing all the options into a single CompoundData isn't ideal, because it makes it impossible for someone to override just one thing in isolation, which thwarts the cascading nature of startup files. Metadata is already a key-value store, so I think we can just use multiple keys instead of a different dictionary.

ericmehl replied:
I got a little overzealous trying to reduce metadata entries. Fixed in TGCCMO2024: https://github.com/GafferHQ/gaffer/pull/6003/commits/8bf9543bafcf26dd41e265f7f18376a2d8460a4a

johnhaddon reviewed (commented):
Thanks Eric!

I didn't manage to spot the code change that removed the spacing between the numeric widgets and the sliders, but I think we should restore the original spacing. Butting up the rounded edge of the numeric field with the flat edge of the slider reads strangely and although it would be fixable by a flat join and a rounded right end of the slider, I think the widgets are distinct enough to deserve their original spacing. It also feels like we could do with a similar spacing on the left of the menu icon? I suspect you're looking for ways of getting a bit more horizontal space now we're squeezing the sliders from both sides, but I think we do need the spacing for clarity still.

My other comments are a mixed bag of API/metadata stuff and UX/bikeshedding stuff. For your sanity, it might be best to attack the former first as I think it's reasonably objective. The rest is undoubtedly subjective, and should be weighed up against other opinions, and - i think - some testing with real users.

Cheers...
John

ericmehl commented on python/GafferUI/ColorChooser.py:
```diff
@@ -163,16 +163,32 @@ def _displayTransformChanged( self ) :
+
+		self.__size = 216
+
+	def resizeEvent( self, event ) :
+
+		w = event.size().width()
+		h = event.size().height()
+
+		if w != h :
+			self.__size = h
+			self.setMinimumWidth( h )
```
Yes, this will be doing a double layout (or maybe even more depending on how the constraints work out?) before `w == h` and it can stop.

The closest I've found to a canonical way of keeping something square in Qt is to use `heightForWidth()` where it returns the input width. We'd need something like `widthForHeight()` since we know the height (the total height of the sliders) and need to figure out the width.

I think the most correct solution would probably be to create a custom layout.

ericmehl replied:
In https://github.com/GafferHQ/gaffer/pull/6003/commits/dea359b3080ab8442e655d29bcc753d24aff43bd, I implemented a custom layout specific to the row for the color field and it's working well.

I tried a couple alternatives to avoid the complexity of a custom layout, but they were not as successful. One was listening for the resize event on the slider container and resizing the color field at that time. But it ran into pretty heavy flickering issues on the first layout.

ericmehl commented on Changes.md:
```diff
@@ -1,7 +1,12 @@
+--------
 
+- ColorChooser :
+  - Add sliders for TMI (temperature, magenta, intensity) color space.
```
Fixed and squashed into their relevant commits.

ericmehl commented on python/GafferUI/ColorChooser.py:
```diff
@@ -548,6 +548,16 @@ def __init__( self, color=imath.Color3f( 1 ), **kw ) :
 							slider.valueChangedSignal().connect( Gaffer.WeakMethod( self.__componentValueChanged ), scoped = False )
 						)
 
+				with GafferUI.ListContainer( GafferUI.ListContainer.Orientation.Vertical, spacing = 4, parenting = { "verticalAlignment" : GafferUI.VerticalAlignment.Top } ) :
```
Fixed in https://github.com/GafferHQ/gaffer/pull/6003/commits/c3407d80a4443a1ed57e19b5f9f7a21f766290c0

ericmehl commented on python/GafferUI/ColorChooser.py:
```diff
@@ -548,6 +548,16 @@ def __init__( self, color=imath.Color3f( 1 ), **kw ) :
+				with GafferUI.ListContainer( GafferUI.ListContainer.Orientation.Vertical, spacing = 4, parenting = { "verticalAlignment" : GafferUI.VerticalAlignment.Top } ) :
+
+					# Options Button
+					menuDefinition = self.__optionsMenuDefinition()
```
That is quite nice, and it is needed for some additional things like `optionsMenuSignal()` that are coming in below. Done in https://github.com/GafferHQ/gaffer/pull/6003/commits/8c1906c3ab5f049e8e43bd9a20beecedf2504d96

ericmehl commented on python/GafferUI/ColorChooser.py:
```diff
@@ -644,11 +650,35 @@ def __optionsMenuDefinition( self ) :
 			}
 		)
 
+		result.append( "/__fieldComponents__", { "divider": True, "label": "Color Field Components" } )
```
Fixed in https://github.com/GafferHQ/gaffer/pull/6003/commits/69377521dfd761d6e4ea39ea3a5bbcadf33b5683

ericmehl commented on resources/graphics.py:
```diff
@@ -476,14 +476,17 @@
-				"colorChooserStaticChannelHighlightedIcon",
+				"colorChooserStaticHighlightedIcon",
+				"colorChooserStaticCenterIcon",
+                "colorChooserStaticTopIcon",
```
I made them darker in https://github.com/GafferHQ/gaffer/pull/6003/commits/eee59e6526e2517a5c9172a77f83b5a12c70061e. They might be a bit hard to see now, but let's work out whether we even want them and get some user feedback too.

ericmehl commented on resources/graphics.py:
```diff
@@ -476,14 +476,17 @@
+				"colorChooserStaticCenterIcon",
+                "colorChooserStaticTopIcon",
+                "colorChooserStaticBottomIcon",
+                "colorChooserAxisCenterIcon",
```
Fixed in https://github.com/GafferHQ/gaffer/pull/6003/commits/a0ce5d3b88ee61ce78eaac7f8cbe90db65d9ceb9

ericmehl commented:
I pushed a stack of fixups to address a good portion of the comments this afternoon. I also added a small fix to deactivate the static component menu items if the color field is not visible : https://github.com/GafferHQ/gaffer/pull/6003/commits/4bce89b6436e87165e72e37aab23d40755d779f9

I think the main things were down to is improving the layout for keeping the color field square, and generally getting user feedback on the changes overall and especially the choice mechanism of the static channel.

ericmehl commented:
I updated the commit links to the fixups above and added a custom layout for the color chooser row to address the double-layout processing we were doing.

I also added a few commits to address user feedback :

- https://github.com/GafferHQ/gaffer/pull/6003/commits/b3082aae2d49769c62d9c5911705324da15fb632 : Added "axes" to the menu to better explain what the component combination are affecting.
- https://github.com/GafferHQ/gaffer/pull/6003/commits/87e721dc2f784dcfbd2f589109a67e71e5468ebc and https://github.com/GafferHQ/gaffer/pull/6003/commits/ce2d8fd0af564ad89ea757da40a0f40c67d663df : Added the ability to turn off RGB sliders (but not A which is always on).
- https://github.com/GafferHQ/gaffer/pull/6003/commits/6aa17ea3fe1e6d308d450896b5c0693a45278d8b : Added a small space between the RGB, A, HSV and TMI channel groups.
- https://github.com/GafferHQ/gaffer/pull/6003/commits/6aa17ea3fe1e6d308d450896b5c0693a45278d8b : Changed the interaction for picking which components are in the color field. You now click on a color you want to be in the field, and it will pop up a menu allowing you to choose from the combinations that use that color.
- https://github.com/GafferHQ/gaffer/pull/6003/commits/6aa17ea3fe1e6d308d450896b5c0693a45278d8b : Added a tooltip to the component label to hint at how to change the components in the color field.
- https://github.com/GafferHQ/gaffer/pull/6003/commits/2c92c30adb485f8a0789fd491a7960d6f229efbf : Changed the wording for saving defaults to emphasize that it's saving the layout, not the color values, as default.

I think that takes care of the code comments and most of the user-feedback for this PR. Would be great to get some interactive testing on the what-components-are-in-the-color-field interaction in particular.

Ready for a new look!

murraystevenson commented on python/GafferUI/ColorChooser.py:
```diff
@@ -454,6 +451,121 @@ def __paintEvent( self, event ) :
+		# line up against the color field. Do add a little space between the sliders
+		# and the options button.
+		self.setSpacing( 0 )
+		self.__buttonSpacing = 4
```
Super nitpicks inbound, but I do feel we've lost a little something in the clarity of the slider layout compared to what we have in current releases.

- We've lost our usual amount of spacing between the NumericWidget and the Slider.
- When the colour field is disabled, we still reserve a bunch of left padding in the component labels for colour field component indicators that will never be shown. This is mostly evident in the inline chooser, where we now indent the labels a fair amount compared to their positions in current Gaffer releases without the colour field.
- Ideally, the inline chooser would align the right edges of the sliders with the right edge of the colour swatch above them, and align the options button with the show/hide inline chooser button above.

I'm happy to punt on bullet point 3, but feel the others are worth considering.

murraystevenson reviewed (commented):
Thanks Eric! Many apologies in advance for continuing the bikeshedding with a few inline UX-related comments after playing with these latest changes...

ericmehl commented on python/GafferUI/ColorChooser.py:
```diff
@@ -474,29 +586,35 @@ def __init__( self, color=imath.Color3f( 1 ), **kw ) :
+						row = { "r" : 0, "g" : 1, "b" : 2, "a" : 4, "h" : 6, "s" : 7, "v" : 8, "t" : 10, "m" : 11, "i" : 12 }[component]
+						self.__channelLabels[component] = GafferUI.Label(
+							component.capitalize(),
+							toolTip = "Click to use this component in the color field.",
```
Right! That shouldn't be on "A". Fixed in https://github.com/GafferHQ/gaffer/pull/6003/commits/20d9b76c436aeeed5196bc9390c6bced2facd2be

ericmehl commented on python/GafferUI/ColorChooser.py:
```diff
@@ -606,6 +807,91 @@ def changesShouldBeMerged( cls, firstReason, secondReason ) :
+			menuDefinition.append(
+				"/" + "".join( sorted( components, key = lambda c : "RGBHSVTMI".index( c ) ) ),
+				{
+					"command": functools.partial( Gaffer.WeakMethod( self.setColorFieldStaticComponent ), yCandidates[ 1 - index ] ),
```
I added a check box next to the active configuration in https://github.com/GafferHQ/gaffer/pull/6003/commits/231b99e51b9686a675aae223611e996332b4f357 to improve this.

ericmehl commented on python/GafferUI/ColorChooser.py:
```diff
@@ -513,6 +631,11 @@ def __init__( self, color=imath.Color3f( 1 ), **kw ) :
 							)
 
+						spacerHeight = 4
+						self.__spacers["a"] = GafferUI.Spacer( imath.V2i( 0, spacerHeight ), parenting = { "index" : ( 0, 3 ) } )
```
There wasn't any user feedback either way on where alpha should go, just that it would be helpful to have a small spacer between triplet groups. I'm separating alpha from RGB since you can now toggle the RGB group on and off, but not alpha (we decided alpha is it's own concept, and allowing it to be toggled would raise confusion with the `Color3f` vs `Color4f` plugs).

I don't have a real strong opinion on it - I could be convinced if you think it's better being grouped with RGB.

murraystevenson replied:
It might just be the spacer height producing a lot of empty space around the A that is throwing me. At present, the spacing between slider groups is greater than the spacing we maintain between plugs in the Node Editor, which might be a bit much? Dropping `spacerHeight` down to 2 could be a happy medium.

ericmehl replied:
In 2dde05b96b34791b51da863c248924c14ffaba8e, I made it 2px spacing, give that a look and see if it feels better.

ericmehl commented:
I pushed a raft of new commits that address the comments above, which I've noted inline. There are a couple of additional commits too :
- https://github.com/GafferHQ/gaffer/pull/6003/commits/d30e3a45921cc1e25e2a27ee84a4fc8cf99a5d36 : Prevents the popup menu for the color field choices from appearing if the color field is not visible.
- https://github.com/GafferHQ/gaffer/pull/6003/commits/931eae170071e6e4c559bc1ba59247f960b220cd : Fixes the tests to be compatible with the new lazy-create of the inline color chooser from https://github.com/GafferHQ/gaffer/pull/6063

murraystevenson commented on .github/workflows/main/installArnold.py:
```diff
@@ -68,6 +69,11 @@
 
 print( "Downloading Arnold \"{}\"".format( url ) )
 
+context = ssl.create_default_context()
```
Seems we might have a rogue commit here?

ericmehl replied:
Ah yes, that was left over from when I needed to push it to an Arnold branch for user testing. Removed in the latest push along with a couple others that were superfluous (fixing `.gfr` whitespace and a cherry pick from #5997).

murraystevenson commented on python/GafferUI/ColorChooserPlugValueWidget.py:
```diff
@@ -103,3 +171,49 @@ def __allComponentsEditable( self ) :
+
+def saveDefaultOptions( colorChooser, keyPrefix, scriptPath = None ) :
+
+	for k in [ "visibleComponents", "staticComponent", "colorFieldVisible" ] :
```
Just a passing thought, but any preference to looping over the plug types here rather than duplicating the metadata deregister/register calls?

ericmehl replied:
Yes, that's a bit cleaner now, in https://github.com/GafferHQ/gaffer/pull/6003/commits/c09ac5960ab3f2762c46b33f31a9611448ce0362

murraystevenson reviewed (changes requested):
Thanks Eric! Feels like we're pretty close to something we can ship in a 1.5 alpha. Few comments inline but nothing too major...

murraystevenson commented on python/GafferUI/ColorChooser.py:
```diff
@@ -454,6 +451,121 @@ def __paintEvent( self, event ) :
+
+		leftSize = 0
+		if not self.__items[0].isEmpty() :
+			self.__items[0].setGeometry( QtCore.QRect( rect.left(), rect.top(), size1.height(), size1.height() ) )
```
Also happy to leave this to a follow-up PR after we get this and #5999 merged.

ericmehl replied:
I was a little shy of doing this at first because as you mentioned it somewhat opens the door to more configurability, but the heuristic works pretty well and is nice and simple so it seems worth doing. Done in https://github.com/GafferHQ/gaffer/pull/6003/commits/f448a562c750d9b3506ff563f046630129763832

ericmehl commented on python/GafferUI/ColorChooser.py:
```diff
@@ -594,6 +810,36 @@ def colorChangedSignal( self ) :
+
+		return self.__visibleComponentsChangedSignal
+
+	## A signal emitted whenver the static component is changed. Slots
```
Fixed in 08a81ec2c13586bd5b5111742764e913f17ca893

ericmehl commented on python/GafferUI/ColorChooser.py:
```diff
@@ -474,29 +586,55 @@ def __init__( self, color=imath.Color3f( 1 ), **kw ) :
+				with GafferUI.GridContainer( spacing = 0 ) :
 
 					# sliders and numeric widgets
 					c, staticComponent = self.__colorField.getColor()
```
Fixed in https://github.com/GafferHQ/gaffer/pull/6003/commits/7fe4f9d461a6202dc4d21f20557307cd8f2d6b4b

ericmehl commented on Changes.md:
```diff
@@ -242,6 +242,14 @@ Features
+- ColorChooser :
+  - Added sliders for TMI (temperature, magenta, intensity) color space.
+  - Added color field, a widget giving control of two channels of "RGB", "HSV" or "TMI" triplets. The third channel is held constant.
+  - The color field, HSV slider group and TMI slider group can now be toggled on or off. Visibility of the elements is maintained for the duration of the Gaffer session.
```
Fixed in the original commits for these entries.

ericmehl commented on Changes.md:
```diff
@@ -242,6 +242,14 @@ Features
+  - Default visibility of the UI elements can now be set at startup by setting `colorChooser:inlineOptions` and `colorChooser:dialogueOptions` `userData` entries for the inline chooser and dialogue chooser respectively. The value of the metadata entry is an `IECore.CompoundData` object with the following optional keys :
+    - `visibleComponents` : A string where each character is a visible component slider. Optional components are `rgbahsvtmi`.
+    - `staticComponent` : A single character string for the component to use as the static component for the color field.
+    - `colorFieldVisible` : A boolean indicating if the color field should be visible or not.
```
Fixed in the original commits for these entries.

ericmehl commented:
All the latest notes are addressed in my most recent push. Hopefully very close and I can do a big squash!

murraystevenson commented on Changes.md:
```diff
@@ -5,6 +5,15 @@ Features
+  - Added sliders for TMI (temperature, magenta, intensity) color space.
+  - Added color field, a widget giving control of two channels of "RGB", "HSV" or "TMI" triplets. The third channel is held constant.
+  - The color field and RGB, HSV and TMI slider groups can now be toggled on or off.
+  - Default visibility of the UI elements can now be set at startup by setting `colorChooser:inlineOptions:` and `colorChooser:dialogueOptions:` entries for the inline chooser and dialogue chooser respectively, with the following suffixes :
```
I think we should be referring to `colorChooser:inline:` and `colorChooser:dialogue:` prefixed metadata here?

ericmehl replied:
Fixed and squashed into https://github.com/GafferHQ/gaffer/pull/6003/commits/19eb8a51a773e65d0ea3afc9271c6060bdfee1f0.

murraystevenson reviewed (commented):
Thanks Eric! One last comment about the changelog, but feel free to address it while you're squashing everything down.

ericmehl commented:
I squashed everything down today and went through and edited the commits to make sure they make sense in isolation and do what they claim to do.

The only change I made was removing a few entries left over in `_StyleSheet` that aren't needed for the current way of setting the icons indicating which components are in the color field.

Checking against a backup branch from before all the rebasing shows only changes to the `_StyleSheet` as expected. After a round of interactive testing, I'm feeling good about this being ready to go!

murraystevenson commented:
Thanks! Merging!
